OpenOversight/tests/test_commands.py

```diff
@@ -79,66 +106,639 @@ def test_add_job_title__success(session):
+    assert n_created > 0
+    assert n_updated == 0
+    officer_count = Officer.query.count()
+    assert officer_count == n_created
+
+    n_created, n_updated = bulk_add_officers([csvfile], standalone_mode=False)
+    assert n_created == 0
+    assert n_updated == 0
+    assert Officer.query.count() == officer_count
+
+
+def test_csv_missing_required_field(csvfile):
+    df = pd.read_csv(csvfile)
+    df.drop(columns="first_name").to_csv(csvfile)
+
+    with pytest.raises(Exception) as exc:
+        bulk_add_officers([csvfile])
+    assert "Missing required field" in str(exc.value)
+
+
+def test_csv_missing_badge_and_uid(csvfile):
+    df = pd.read_csv(csvfile)
+    df.drop(columns=["star_no", "unique_internal_identifier"]).to_csv(csvfile)
+
+    with pytest.raises(Exception) as exc:
+        bulk_add_officers([csvfile])
+    assert (
+        "CSV file must include either badge numbers or unique identifiers for officers"
+        in str(exc.value)
+    )
+
+
+def test_csv_non_existant_dept_id(csvfile):
+    df = pd.read_csv(csvfile)
+    df["department_id"] = 666
+    df.to_csv(csvfile)
+
+    with pytest.raises(Exception) as exc:
+        bulk_add_officers([csvfile])
+    assert "Department ID 666 not found" in str(exc.value)
+
+
+def test_csv_officer_missing_badge_and_uid(csvfile):
+    df = pd.read_csv(csvfile)
+    df.loc[0, "star_no"] = None
+    df.loc[0, "unique_internal_identifier"] = None
+    df.to_csv(csvfile)
+
+    with pytest.raises(Exception) as exc:
+        bulk_add_officers([csvfile])
+    assert "missing badge number and unique identifier" in str(exc.value)
+
+
+def test_csv_changed_static_field(csvfile):
+    df = pd.read_csv(csvfile)
+    df.loc[0, "birth_year"] = 666
+    df.to_csv(csvfile)
+
+    with pytest.raises(Exception) as exc:
+        bulk_add_officers([csvfile])
+    assert "has differing birth_year field" in str(exc.value)
+
+
+def test_csv_new_assignment(csvfile):
+    # Delete all current officers and assignments
+    Assignment.query.delete()
+    Officer.query.delete()
+
+    assert Officer.query.count() == 0
+
+    df = pd.read_csv(csvfile)
+    df.loc[0, "job_title"] = "Commander"
+    df.to_csv(csvfile)
+
+    n_created, n_updated = bulk_add_officers([csvfile], standalone_mode=False)
+    assert n_created > 0
+    assert n_updated == 0
+    assert Officer.query.count() == n_created
+
+    officer = get_officer(
+        1, df.loc[0, "star_no"], df.loc[0, "first_name"], df.loc[0, "last_name"]
+    )
+    assert officer
+    officer_id = officer.id
+    assert len(list(officer.assignments)) == 1
+
+    # Update job_title
+    df.loc[0, "job_title"] = "CAPTAIN"
+    df.to_csv(csvfile)
+
+    n_created, n_updated = bulk_add_officers([csvfile], standalone_mode=False)
+    assert n_created == 0
+    assert n_updated == 1
+
+    officer = Officer.query.filter_by(id=officer_id).one()
+    assert len(list(officer.assignments)) == 2
+    for assignment in officer.assignments:
```

Comment:
I see what you are doing w/ this loop, but it's not strict enough. If both assignments were "CAPTAIN", or both assignments were "Commander", the test would pass. It should check that there's one of each.

Reply:
Side note, I may have a pull request coming soon that would be relevant to this. But, this test can stay as-is (aside from the comment about the for loop above), I'll submit a new test along with it. I will make it a separate PR so that we can get this one finished with.

Reply:
Turns out this is an issue with code that was already in the codebase before this PR. It is showing up in the commit because the code was moved into a new file. I am marking this issue "resolved." @abandoned-prototype has indicated in Slack he might submit a separate PR for this next week. I also am happy to do it.